File: example/android/src/main/kotlin/com/cmc/c_shield_sdk/streams/ThreatEventStreamHandler.kt
package com.cmc.c_shield_sdk.streams

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/**
 * Bridges [com.example.c_shield_sdk.LoadAppThreatListener] callbacks to Dart.
 *
 * Best-effort by design: the native loader tears the process down shortly after
 * the callback returns, so Dart may or may not get to act on the event. Use it
 * for local logging and diagnostics, never as the mechanism that displays the
 * threat popup — that is the native side's job precisely because it is the only
 * layer guaranteed to still be alive.
 *
 * Load-time threats can also fire before Dart has subscribed; events with no
 * sink are buffered and replayed on the next [onListen].
 */
class ThreatEventStreamHandler : EventChannel.StreamHandler {

    private val mainHandler = Handler(Looper.getMainLooper())

    // Only touched on the main thread.
    private var sink: EventChannel.EventSink? = null
    private val pending = ArrayDeque<Int>()

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        sink = events
        while (pending.isNotEmpty()) {
            events.success(mapOf("threatType" to pending.removeFirst()))
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    /** Called from a native thread — hops to main before touching the sink. */
    fun emit(threatType: Int) {
        mainHandler.post {
            val s = sink
            if (s == null) pending.addLast(threatType) else s.success(
                mapOf("threatType" to threatType)
            )
        }
    }
}

File: example/ios/Classes/ThreatEventStreamHandler.swift
import Flutter

/// Bridges `CShield` load-time threat notifications to Dart over the
/// `c_shield_sdk/threat_events` EventChannel. Mirrors Android's
/// `ThreatEventStreamHandler`.
///
/// Best-effort by design: the native SDK kills the process shortly after a
/// load-time threat, so Dart may or may not get to act. For logging only —
/// the popup is drawn natively.
///
/// A threat can also fire before Dart has subscribed (the `sdk.initialize`
/// method call and the EventChannel `onListen` arrive on different channels,
/// whose relative ordering is not guaranteed). Events emitted with no sink are
/// buffered and replayed on the next `onListen`.
final class ThreatEventStreamHandler: NSObject, FlutterStreamHandler {

    private var sink: FlutterEventSink?
    private var pending: [Int] = []

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        for threatType in pending { events(["threatType": threatType]) }
        pending.removeAll()
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }

    /// May be called from any thread (the native listener runs on whichever
    /// thread invoked `CShield.initialize`); hops to main before touching the
    /// sink, which Flutter requires.
    func emit(_ threatType: Int) {
        if Thread.isMainThread {
            deliver(threatType)
        } else {
            DispatchQueue.main.async { [weak self] in self?.deliver(threatType) }
        }
    }

    private func deliver(_ threatType: Int) {
        if let sink = sink {
            sink(["threatType": threatType])
        } else {
            pending.append(threatType)
        }
    }
}
